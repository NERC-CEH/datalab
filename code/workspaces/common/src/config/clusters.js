import data from './clusters_config.json';

const clustersConfig = () => data;
export default clustersConfig;
