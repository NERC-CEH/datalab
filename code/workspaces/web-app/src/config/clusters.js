import axios from 'axios';

// This is a cached value.
// Initialise asynchronously before React starts.
// Get synchronously within React.
let cachedData;

export async function initialiseClusters() {
  const { data } = await axios.get('/clusters_config.json');
  cachedData = data;
}

export const getClusterMaxWorkers = type => cachedData[type.toLowerCase()].cluster.workersMax;
export const getWorkerMemoryMax = type => cachedData[type.toLowerCase()].workers.memoryMax_GB;
export const getWorkerCpuMax = type => cachedData[type.toLowerCase()].workers.CpuMax_vCPU;
export const getCondaRequired = type => cachedData[type.toLowerCase()].condaRequired;
