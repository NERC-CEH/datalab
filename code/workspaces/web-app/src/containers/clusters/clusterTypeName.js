export const CLUSTER_TYPE_NAME = 'Cluster';
export const CLUSTER_TYPE_NAME_PLURAL = 'Clusters';
export const DASK_CLUSTER_TYPE = 'DASK';
export const SPARK_CLUSTER_TYPE = 'SPARK';
