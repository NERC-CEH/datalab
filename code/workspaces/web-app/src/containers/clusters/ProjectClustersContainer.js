import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { permissionTypes, statusTypes } from 'common';
import { projectKeyPermission } from 'common/src/permissionTypes';
import { reset } from 'redux-form';
import StackCards from '../../components/stacks/StackCards';
import { useClustersByType } from '../../hooks/clustersHooks';
import clusterActions from '../../actions/clusterActions';
import { useCurrentUserId } from '../../hooks/authHooks';
import modalDialogActions from '../../actions/modalDialogActions';
import { getClusterMaxWorkers, getWorkerCpuMax, getWorkerMemoryMax, getCondaRequired } from '../../config/clusters';
import { MODAL_TYPE_CREATE_CLUSTER, MODAL_TYPE_CONFIRMATION } from '../../constants/modaltypes';
import { useDataStorageForUserInProject } from '../../hooks/dataStorageHooks';
import dataStorageActions from '../../actions/dataStorageActions';
import notify from '../../components/common/notify';
import { CLUSTER_TYPE_NAME, CLUSTER_TYPE_NAME_PLURAL } from './clusterTypeName';
import assetRepoActions from '../../actions/assetRepoActions';

const refreshInterval = 15000;

const { projectPermissions: { PROJECT_KEY_CLUSTERS_CREATE, PROJECT_KEY_CLUSTERS_DELETE, PROJECT_KEY_CLUSTERS_EDIT, PROJECT_KEY_CLUSTERS_OPEN } } = permissionTypes;

export const FORM_NAME = 'createCluster';

const deleteCluster = async (dispatch, cluster) => {
  try {
    await dispatch(clusterActions.deleteCluster(cluster));
    dispatch(modalDialogActions.closeModalDialog());
    notify.success('Cluster deleted.');
  } catch (error) {
    notify.error('Unable to delete cluster.');
  } finally {
    dispatch(clusterActions.loadClusters(cluster.projectKey));
  }
};

const confirmDeleteCluster = dispatch => cluster => dispatch(modalDialogActions.openModalDialog(MODAL_TYPE_CONFIRMATION, {
  title: `Delete ${cluster.displayName} cluster`,
  body: `Would you like to delete the ${cluster.displayName} cluster?
  The cluster will stop and be permanently deleted.`,
  onSubmit: () => deleteCluster(dispatch, cluster),
  onCancel: () => dispatch(modalDialogActions.closeModalDialog()),
}));

const scaleCluster = async (dispatch, cluster, replicas) => {
  try {
    dispatch(modalDialogActions.closeModalDialog());
    await dispatch(clusterActions.scaleCluster(cluster, replicas));
    notify.success(`Cluster ${replicas > 0 ? 'started' : 'suspended'}`);
  } catch (error) {
    notify.error('Unable to scale cluster');
  } finally {
    dispatch(clusterActions.loadClusters(cluster.projectKey));
  }
};

export const confirmScaleCluster = dispatch => async (cluster) => {
  if (cluster.status === statusTypes.SUSPENDED) {
    scaleCluster(dispatch, cluster, 1);
    return;
  }

  dispatch(modalDialogActions.openModalDialog(MODAL_TYPE_CONFIRMATION, {
    title: `Suspend ${cluster.displayName} cluster?`,
    body: `Would you like to suspend the ${cluster.displayName} cluster?`,
    onSubmit: () => scaleCluster(dispatch, cluster, 0),
    onCancel: () => dispatch(modalDialogActions.closeModalDialog()),
  }));
};

const getOpenCreationForm = (dispatch, projectKey, clusterType, dataStores) => {
  if (clusterType) {
    const createClusterDialogProps = getDialogProps(dispatch, projectKey, clusterType, dataStores);
    return () => dispatch(modalDialogActions.openModalDialog(MODAL_TYPE_CREATE_CLUSTER, createClusterDialogProps));
  }

  return undefined;
};

const ProjectClustersContainer = ({ clusterType, projectKey, userPermissions, modifyData, copySnippets }) => {
  const dispatch = useDispatch();
  const currentUserId = useCurrentUserId();
  const { value: dataStores } = useDataStorageForUserInProject(currentUserId, projectKey);
  const clusters = clustersToStacks(useClustersByType(clusterType), projectKey);

  useEffect(() => {
    if (projectKey && modifyData) {
      dispatch(clusterActions.loadClusters(projectKey));
      dispatch(dataStorageActions.loadDataStorage(projectKey));
      dispatch(assetRepoActions.loadVisibleAssets(projectKey));
    }

    // After an interval, update the clusters list in case of any status changes.
    // This uses "updateClusters" to avoid flickering.
    const interval = setInterval(() => {
      if (projectKey && modifyData) {
        dispatch(clusterActions.updateClusters(projectKey));
        dispatch(dataStorageActions.loadDataStorage(projectKey));
        dispatch(assetRepoActions.loadVisibleAssets(projectKey));
      }
    }, refreshInterval);

    return () => clearInterval(interval);
  }, [dispatch, modifyData, projectKey]);

  const openCreationForm = getOpenCreationForm(dispatch, projectKey, clusterType, dataStores);

  return (
    <StackCards
      stacks={clusters}
      typeName={CLUSTER_TYPE_NAME}
      typeNamePlural={CLUSTER_TYPE_NAME_PLURAL}
      userPermissions={() => userPermissions}
      openCreationForm={openCreationForm}
      createPermission={projectKeyPermission(PROJECT_KEY_CLUSTERS_CREATE, projectKey)}
      showCreateButton={modifyData}
      deleteStack={modifyData ? confirmDeleteCluster(dispatch) : undefined}
      scaleStack={modifyData ? confirmScaleCluster(dispatch) : undefined}
      copySnippets={copySnippets}
      deletePermission={projectKeyPermission(PROJECT_KEY_CLUSTERS_DELETE, projectKey)}
      editPermission={projectKeyPermission(PROJECT_KEY_CLUSTERS_EDIT, projectKey)}
      openPermission={projectKeyPermission(PROJECT_KEY_CLUSTERS_OPEN, projectKey)}
    />
  );
};

const clustersToStacks = (clusters, projectKey) => ({
  ...clusters,
  value: clusters.value
    .filter(cluster => cluster.projectKey === projectKey)
    .map(cluster => ({
      ...cluster,
      description: `${cluster.schedulerAddress} is the scheduler address.`,
    })),
});

export const getDialogProps = (dispatch, projectKey, clusterType, dataStores) => ({
  title: `Create ${CLUSTER_TYPE_NAME}`,
  formName: FORM_NAME,
  onSubmit: async (cluster) => {
    dispatch(modalDialogActions.closeModalDialog());
    let creationSuccess;
    try {
      await dispatch(clusterActions.createCluster({
        type: clusterType,
        projectKey,
        ...cluster,
      }));
      creationSuccess = true;
    } catch (error) {
      creationSuccess = false;
    }
    if (creationSuccess) {
      notify.success(`${CLUSTER_TYPE_NAME} created`);
      dispatch(reset(FORM_NAME));
      dispatch(clusterActions.loadClusters(projectKey));
    } else {
      notify.error(`Unable to create ${CLUSTER_TYPE_NAME}`);
    }
  },
  onCancel: () => dispatch(modalDialogActions.closeModalDialog()),
  clusterMaxWorkers: getClusterMaxWorkers(clusterType),
  workerMaxMemory: getWorkerMemoryMax(clusterType),
  workerMaxCpu: getWorkerCpuMax(clusterType),
  condaRequired: getCondaRequired(clusterType),
  dataStorageOptions: dataStores.map(store => ({ text: store.displayName, value: store.name })),
  projectKey,
});

export default ProjectClustersContainer;

ProjectClustersContainer.propTypes = {
  clusterType: PropTypes.string,
  copySnippets: PropTypes.objectOf(PropTypes.func),
};
