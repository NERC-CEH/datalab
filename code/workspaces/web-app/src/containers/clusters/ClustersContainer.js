import PropTypes from 'prop-types';
import { useCurrentProjectKey } from '../../hooks/currentProjectHooks';
import { useCurrentUserPermissions } from '../../hooks/authHooks';
import ProjectClustersContainer from './ProjectClustersContainer';

const ClustersContainer = ({ clusterType, copySnippet }) => {
  const { value: projectKey } = useCurrentProjectKey();
  const { value: userPermissions } = useCurrentUserPermissions();
  const modifyData = true;
  return ProjectClustersContainer({ clusterType, projectKey, userPermissions, modifyData, copySnippet });
};

export default ClustersContainer;

ClustersContainer.propTypes = {
  clusterType: PropTypes.string.isRequired,
  copySnippet: PropTypes.func,
};
