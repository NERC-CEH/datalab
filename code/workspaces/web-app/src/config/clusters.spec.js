import MockAdapter from 'axios-mock-adapter';
import axios from 'axios';
import data from 'common/src/config/clusters_config.json';
import { initialiseClusters, getClusterMaxWorkers, getWorkerCpuMax, getWorkerMemoryMax, getCondaRequired } from './clusters';

const httpMock = new MockAdapter(axios);
httpMock.onGet('/clusters_config.json').reply(() => [200, data]);

const CLUSTER_TYPE = 'DASK';

describe('getClusterMaxWorkers', () => {
  it('returns correct value', async () => {
    await initialiseClusters();
    const maxWorkers = getClusterMaxWorkers(CLUSTER_TYPE);
    expect(maxWorkers).toMatchSnapshot();
  });
});

describe('getWorkerMemoryMax', () => {
  it('returns correct value', async () => {
    await initialiseClusters();
    const maxMemory = getWorkerMemoryMax(CLUSTER_TYPE);
    expect(maxMemory).toMatchSnapshot();
  });
});

describe('getWorkerCpuMax', () => {
  it('returns correct value', async () => {
    await initialiseClusters();
    const maxCpu = getWorkerCpuMax(CLUSTER_TYPE);
    expect(maxCpu).toMatchSnapshot();
  });
});

describe('getCondaRequired', () => {
  it('returns correct value', async () => {
    await initialiseClusters();
    const condaRequired = getCondaRequired(CLUSTER_TYPE);
    expect(condaRequired).toBeFalsy();
  });
});

