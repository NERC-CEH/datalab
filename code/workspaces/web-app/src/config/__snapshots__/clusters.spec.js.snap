// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getClusterMaxWorkers returns correct value 1`] = `
Object {
  "default": 4,
  "lowerLimit": 1,
  "upperLimit": 8,
}
`;

exports[`getWorkerCpuMax returns correct value 1`] = `
Object {
  "default": 0.5,
  "lowerLimit": 0.5,
  "upperLimit": 2,
}
`;

exports[`getWorkerMemoryMax returns correct value 1`] = `
Object {
  "default": 4,
  "lowerLimit": 0.5,
  "upperLimit": 8,
}
`;
